cdap-app-fabric/src/main/java/co/cask/cdap/scheduler/ProgramScheduleService.java

```diff
@@ -16,84 +16,227 @@
+   * @throws Exception if any other errors occurred while performing the authorization enforcement check
    */
-  public ProgramScheduleStatus getScheduleStatus(ScheduleId scheduleId)
-    throws Exception {
-    ApplicationId applicationId = scheduleId.getParent();
-    ApplicationSpecification appSpec = store.getApplication(applicationId);
-    if (appSpec == null) {
-      throw new NotFoundException(applicationId);
+  public void update(ScheduleId scheduleId, ScheduleDetail scheduleDetail) throws Exception {
+    authorizationEnforcer.enforce(scheduleId.getParent(), authenticationContext.getPrincipal(), Action.ADMIN);
+    ProgramSchedule existing = scheduler.getSchedule(scheduleId);
+
+    String description = Objects.firstNonNull(scheduleDetail.getDescription(), existing.getDescription());
+    ProgramId programId = scheduleDetail.getProgram() == null ? existing.getProgramId()
+      : existing.getProgramId().getParent().program(
+      scheduleDetail.getProgram().getProgramType() == null ? existing.getProgramId().getType()
+        : ProgramType.valueOfSchedulableType(scheduleDetail.getProgram().getProgramType()),
+      Objects.firstNonNull(scheduleDetail.getProgram().getProgramName(), existing.getProgramId().getProgram()));
+    if (!programId.equals(existing.getProgramId())) {
+      throw new BadRequestException(
+        String.format("Must update the schedule '%s' with the same program as '%s'. "
+                        + "To change the program in a schedule, please delete the schedule and create a new one.",
+                      existing.getName(), existing.getProgramId().toString()));
     }
+    Map<String, String> properties = Objects.firstNonNull(scheduleDetail.getProperties(), existing.getProperties());
+    Trigger trigger = Objects.firstNonNull(scheduleDetail.getTrigger(), existing.getTrigger());
+    List<? extends Constraint> constraints =
+      Objects.firstNonNull(scheduleDetail.getConstraints(), existing.getConstraints());
+    Long timeoutMillis = Objects.firstNonNull(scheduleDetail.getTimeoutMillis(), existing.getTimeoutMillis());
+    ProgramSchedule updatedSchedule = new ProgramSchedule(existing.getName(), description, programId, properties,
+                                                          trigger, constraints, timeoutMillis);
+    scheduler.updateSchedule(updatedSchedule);
+  }
+
+  /**
+   * Get the given schedule
+   *
+   * @return the schedule
+   * @throws NotFoundException if the schedule could not be found
+   * @throws UnauthorizedException if the principal is not authorized to access the schedule program
+   * @throws Exception if any other errors occurred while performing the authorization enforcement check
+   */
+  public ProgramSchedule get(ScheduleId scheduleId) throws Exception {
+    ProgramSchedule schedule = scheduler.getSchedule(scheduleId);
+    AuthorizationUtil.ensureAccess(schedule.getProgramId(), authorizationEnforcer,
```

Comment:
better to check the authorization at the application level with `schedule.getParent()` to be consistent with methods like add and update

Reply:
these are different things, parent of a schedule is an app, this is a program. I just took whatever logic was currently there. Is it wrong logic? I couldn't find any docs on how schedule authorization is supposed to be.